backend/services/job.js

```diff
@@ -0,0 +1,125 @@
+/**
+ * JobService that interacts with the job documents in the database
+ */
+import { CustomError, InternalError, ServiceError } from '../errors';
+import JobModel, { FIELDS_OWNER_PERMITTED_TO_UPDATE } from '../models/job';
+import { saveImage, deleteImage } from './image';
+
+export async function createJob(user, jobData, jobImages) {
+    console.debug('createJob service running')
+    const imageIds = [];
+
+    for (let image of jobImages) {
+        let imageId = await saveImage(image);
+        imageIds.push(imageId);
+    }
+
+    let job = null;
+    try {
+        job = await JobModel({
+            ...jobData, //TODO Sanitize input
+            client: user._id,
+            applicants: [],
+            imageIds: imageIds,
+        });
+    } catch (e) {
+        throw ServiceError.INVALID_JOB_RECEIVED.addContext(e.stack);
+    }
+    
+    //TODO Validate request fields like delievryDate, etc.
+
+    try {
+        return await job.save();
+    } catch (e) {
+        throw ServiceError.INVALID_JOB_RECEIVED.addContext(e.stack);
+    }
+}
+
+function filterUpdatePayload(payload, allowedFields) {
+    const filtered = {}  
+    allowedFields.forEach(field => {
+      if (field in payload) filtered[field] = payload[field];
+    });
+    return filtered;
+}
+
+export async function updateJob(user, jobId, jobData, jobImages) {
+    console.debug('updateJob service runnning');
+
+    // Retrieve original job
+    let originalJob = await JobModel.findById(jobId);
+    if (!originalJob) {
+        throw ServiceError.JOB_NOT_FOUND;
+    }
+
+    // Validate client job ownership
+    if (!originalJob.client.equals(user._id)) {
+        throw ServiceError.JOB_EDIT_PERMISSION_DENIED;
+    }
+
+    // Ensure updated fields are only getting updated
+    jobData = filterUpdatePayload(jobData, FIELDS_OWNER_PERMITTED_TO_UPDATE);
+
+    //TODO Find a better way of updating images
```

Comment:
Currently, when the update form page is submitted this endpoint should be triggered which will cause the service to remove the existing images and store the images from the update request (irrespective of whether they were changed) into the DB. To improve on this option, we might have to somehow track whether the images where changed on the frontend but not sure if that is feasible.